File: src/main/java/carelog/carelog/journal/domain/RelationJournal.java
package carelog.carelog.journal.domain;

import carelog.carelog.common.domain.TenantBaseEntity;
import carelog.carelog.relation.domain.Relation;
import io.hypersistence.utils.hibernate.type.json.JsonBinaryType;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.Type;

import java.util.Map;
import java.util.UUID;

@Entity
@Table(name = "relation_journals")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class RelationJournal extends TenantBaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "relation_journals_id", updatable = false, nullable = false)
    private Long id;

    @Column(name = "public_id", unique = true, nullable = false, updatable = false)
    private UUID publicId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "relation_id", nullable = false)
    private Relation relation;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "template_id")
    private JournalTemplate template;

    @Type(JsonBinaryType.class)
    @Column(name = "content", columnDefinition = "jsonb", nullable = false)
    private Map<String, Object> content;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    private JournalStatus status;

    @Column(name = "previous_id")
    private Long previousId;


    public RelationJournal(
            Relation relation, JournalTemplate template,
            Map<String, Object> content, Long previousId
    ) {
        this.publicId = UUID.randomUUID();
        this.relation = relation;
        this.template = template;
        this.content = content;
        this.status = JournalStatus.ACTIVE;
        this.previousId = previousId;
    }

    // 최초 진료 일지 생성
    public static RelationJournal create(
            Relation relation, JournalTemplate template,
            Map<String, Object> content
    ) {
        return new RelationJournal(relation, template, content, null);
    }

    // 기존 일지의 수정본 생성 (이전 버전 id 연결)
    public static RelationJournal createAsRevision(
            Relation relation, JournalTemplate template,
            Map<String, Object> content, Long previousId
    ) {
        return new RelationJournal(relation, template, content, previousId);
    }

    // 이전 버전을 대체됨 상태로 변경
    public void supersede() {
        this.status = JournalStatus.SUPERSEDED;
    }

}

File: src/main/kotlin/carelog/carelog/journal/domain/RelationJournal.kt
package carelog.carelog.journal.domain

import carelog.carelog.common.domain.TenantBaseEntity
import carelog.carelog.relation.domain.Relation
import io.hypersistence.utils.hibernate.type.json.JsonBinaryType
import jakarta.persistence.*
import org.hibernate.annotations.Type
import java.time.LocalDate
import java.util.UUID

@Entity
@Table(name = "relation_journals")
class RelationJournal private constructor(
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "relation_id", nullable = false)
    val relation: Relation,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "template_id")
    val template: JournalTemplate?,

    @Column(name = "title", nullable = false)
    val title: String,

    @Column(name = "visit_date", nullable = false)
    val visitDate: LocalDate,

    @Type(JsonBinaryType::class)
    @Column(name = "case_data", columnDefinition = "jsonb", nullable = false)
    val caseData: Map<String, Any>,

    @Type(JsonBinaryType::class)
    @Column(name = "private_data", columnDefinition = "jsonb")
    val privateData: Map<String, Any>?,

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    var status: JournalStatus,

    @Column(name = "previous_id")
    val previousId: Long?,
) : TenantBaseEntity() {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "relation_journals_id", updatable = false, nullable = false)
    val id: Long = 0

    @Column(name = "public_id", unique = true, nullable = false, updatable = false)
    val publicId: UUID = UUID.randomUUID()

    fun supersede() { status = JournalStatus.SUPERSEDED }

    companion object {
        fun create(
            relation: Relation, template: JournalTemplate?,
            title: String, visitDate: LocalDate,
            caseData: Map<String, Any>, privateData: Map<String, Any>?,
        ) = RelationJournal(relation, template, title, visitDate, caseData, privateData, JournalStatus.ACTIVE, null)

        fun createAsRevision(
            relation: Relation, template: JournalTemplate?,
            title: String, visitDate: LocalDate,
            caseData: Map<String, Any>, privateData: Map<String, Any>?,
            previousId: Long,
        ) = RelationJournal(relation, template, title, visitDate, caseData, privateData, JournalStatus.ACTIVE, previousId)
    }
}
